packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -824,9 +825,46 @@ class _TextSelectionHandleOverlayState
         break;
     }
 
+    // On some platforms we may want to calculate the start and end handles
+    // separately so they scale for the selected content.
+    //
+    // For the start handle we compute the rectangles that encompass the range
+    // of the first full selected grapheme cluster at the beginning of the selection.
+    //
+    // For the end handle we compute the rectangles that encompass the range
+    // of the last full selected grapheme cluster at the end of the selection.
+    final InlineSpan span = widget.renderObject.text!;
+    final String text = span.toPlainText();
+    final int firstSelectedGraphemeExtent;
+    final int lastSelectedGraphemeExtent;
+    if (text.isEmpty || widget.renderObject.selection == null || !widget.renderObject.selection!.isValid || widget.renderObject.selection!.isCollapsed) {
```

Comment:
nit: this expression is getting a bit too long. Maybe make selection a local variable, and that way you can promote it to non-null.

Reply:
if the text is empty the selection is guaranteed to be collapsed/invalid I think?

Reply:
Yeah I think that makes sense if both the selection and text belong to the last frame.